### PR TITLE
New Automl script (again)

### PR DESCRIPTION
**Pull request by unmonoqueteclea**
Now the automl script uses only 1 script and a library.
(I wish I could write code as beautiful as my readme files)

I have tested it removing some random different fields from train,
 validation and test datasets. The problem with the previous version
seems to be fixed.

### CODE REVIEW

File: automl/automl-library/readme.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  from your datasets
+- **Feature Generation**: Functions to create new features from the
+  created unsupervised models
+- **Feature Selection**: Functions to obtain, automatically, the most
+  important fields in a dataset. **Recursive Feature Elimination**
+  algorithm is used.
+- **Model Selection**: Functions to obtain, automatically, the best
+  supervised models and their hyperparameters.
+
+Private functions, that are only used inside the library, starts with
```
Comment (mmerce) [inline automl/automl-library/readme.md]: typo:  start with...

File: automl/automl-script/readme.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+![BigML AutoML Steps](../res/steps.png)
+
+From a train, validation and test datasets, the following tasks will be
+automatically done:
+
+-  `Unsupervised Models Generation`: Creating the following
+  **unsupervised models**: `Cluster`, `Anomaly Detector`, `Association
+  Discovery` (with lift and leverage as search_strategies), `PCA` and
+  `Topic Model` (when there are text models in the dataset).
```
Comment (mmerce) [inline automl/automl-script/readme.md]: Topic Model (for datasets that contain text fields)

File: automl/automl-script/readme.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+-  `Feature Generation`: Using the unsupervised models created
+  previously for scoring the datasets and **generating, automatically,
```
Comment (mmerce) [inline automl/automl-script/readme.md]: previously to append **automatically generated new features** to all the user given datasets.

File: automl/automl-script/readme.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+-  `Model Selection` Using OptiML to find the best models and using
+  the top 3 models to create a `Fusion` model to predict all the test
+  dataset instances. If a validation dataset is given, this script
+  will return an evaluation of the final model too.
+
+
+The **inputs** for the script are:
+
+* `train-dataset`: (string) Dataset id for the train dataset
+  (e.g. dataset/5d272205eba31d61920005cd). If you don't provide a
```
Comment (mmerce) [inline automl/automl-script/readme.md]: If no train dataset is provided, the script will expect an `automl-execution` to be provided and will use its models as starting point.

File: automl/automl-script/readme.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  train dataset, the script will reuse the models created in the
+  provided `automl-execution`.
+* `validation-dataset`: (string) Dataset id for the validation dataset
+  (e.g. dataset/5d272205eba31d61920005cd). You can leave this input
```
Comment (mmerce) [inline automl/automl-script/readme.md]: If empty, no `output-evalution` is returned.

File: automl/automl-script/readme.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  blank and the script won't return an `output-evaluation`.
+* `test-dataset`: (string) Dataset id for the test dataset
+  (e.g. dataset/7j272205eba31d61920005vf). You can leave this input
+  blank and the script won't return an `output-dataset`.
```
Comment (mmerce) [inline automl/automl-script/readme.md]: If empty, no `output-dataset` is returned.

File: automl/automl-script/readme.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+* `automl-exection`: (string) Previous execution of this script, to
+  reuse created executions and models,
+  e.g. execution/5d272205eba31d61920005cd. If you don't provide a
+  `train-dataset` you must provide an `automl-exection`. If both are
```
Comment (mmerce) [inline automl/automl-script/readme.md]: Either `train-dataset` or `automl-execution` should be provided for the script to work. When both are provided, the `automl-execution`  argument is discarded and new resources are generated from the training data.

File: automl/automl-script/readme.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  provided, the script will generate new resources instead reusing
+  them.
+* `excluded-fields`: (list) List of field names to exclude from all
+  the datasets. e.g. ["bmi", "age"]. Empty by default
```
Comment (mmerce) [inline automl/automl-script/readme.md]: List of fields that will be excluded from any dataset before any other process starts, e.g. ...

File: automl/automl-script/readme.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+* `output-fusion`: (fusion-id) Output fusion model
+* `selected-fields`: (list) Selected important field names
+* `unsupervised-models`: (list) List of unsupervised models created
+  from train-dataset
+
+## Usage
+There are two different ways of using this script:
+
+### From a train and a test dataset (and an optional validation dataset)
+In this case, we should pass to the script the `train-dataset` and the
```
Comment (mmerce) [inline automl/automl-script/readme.md]: In this case, the expected inputs for the script are the...

File: automl/automl-script/readme.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+`test-dataset`, and we will leave the `automl-execution` input blank.
```
Comment (mmerce) [inline automl/automl-script/readme.md]: with no `autml-execution` input.

File: automl/automl-script/readme.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+Remember to set previously the correct **objective field** in your
```
Comment (mmerce) [inline automl/automl-script/readme.md]: The **objective field** used in the script will be the one associated to the datasets used, so remember to choose them previously in both the training and validation datasets.

File: automl/automl-script/readme.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+and it will return, at the end of the process, the **output-dataset**
+with the final predictions for the test dataset using a `Fusion` with
+the best models from the created `OptiML`.
+
+If a validation dataset is given, the script will also return an
+`evaluation` of the final `Fusion` model with the
+`validation-dataset`.
+
+### From a test dataset and a previous execution id (and an optional validation dataset)
+If we executed previously the `automl script` with a given train
```
Comment (mmerce) [inline automl/automl-script/readme.md]: If we want to predict new data using the same models created by a previously executed `automl script`, you can use the `automl-execution` parameter (no `train-dataset` parameter needed) associated to that previous execution and provide the test dataset ID in the `test-dataset`parameter.

In this case, some steps...

File: automl/automl-script/readme.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+We can pass to the script the new `test-dataset` and the the id of the
+last execution of the script in the `automl-execution` input. We will
+leave `train-dataset` input blank.
+
+In this case, some steps of the process will be bypassed. The script
+won't generate neither a new `OptiML`, unsupervised model nor a
+**Recursive Feature Selection** execution.
+
+## FAQ
+### Why can't I see in, in the scripts inputs in the BigML's dashboard, dropdowns menus to select the dataset, script, etc?
```
Comment (mmerce) [inline automl/automl-script/readme.md]: typo: in, **in**

File: automl/automl-script/readme.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+All the input parameters of these scripts are optional. So, they use a
+**string** input type instead a `dataset-id` input type,
+`execution-id` input type, etc. Using a **string** input type we can
+provide a **default blank value** if the user don't want to pass one
```
Comment (mmerce) [inline automl/automl-script/readme.md]: **default empty value**

File: automl/automl-script/script.whizzml [hunk 15]
```diff
@@ -0,0 +1,126 @@
+(log-featured "Initial checks")
+(log-info "Checking execution inputs")
+(check-inputs train-dataset validation-dataset test-dataset automl-execution)
+(log-info "Checking excluded fields")
+(check-excluded-fields train-dataset automl-execution excluded-fields)
+(log-info "Checking input dataset fields")
+(check-datasets-fields train-dataset validation-dataset test-dataset)
+
+;; Whether we should resources from a given execution
```
Comment (mmerce) [inline automl/automl-script/script.whizzml]: we should **reuse**

File: automl/automl-script/script.whizzml [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+                    (create-unsupervised "cluster" dataset params true)
+                    (create-unsupervised "anomaly" dataset params true)
+                    (create-unsupervised "pca" dataset params true)
+                    (create-unsupervised "topicmodel" dataset params true)])
+    (remove-false all-models)))
+
+;; Returns a list of unsupervised model from an execution output
+(define (get-unsupervised-models automl-exec)
+  (output-from-exec automl-exec "unsupervised-models"))
+;; List will all the created unsupervised models
```
Comment (mmerce) [inline automl/automl-script/script.whizzml]: List will **contain**?...

File: automl/automl-script/script.whizzml [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+       [train-dataset validation-dataset test-dataset]))
+
+
+(log-featured "Feature Selection")
+(log-info "Obtaining the most important features")
+;; Obtains the list of important fields from
+;; a previous execution
+(define (retrieve-important-fields exec)
+  (when (= (resource-type exec) "execution")
+    (output-from-exec exec "selected-fields")))
```
Comment (mmerce) [inline automl/automl-script/script.whizzml:67]: wouldn't it be good to have some whizzml functions like (execution-output or (execution-result  that returned the output of an execution? That's probably something that we are to use often...

File: automl/automl-script/script.whizzml [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+        topic-fields (feat-gen "topicmodel" "batchtopicdistribution")
+        pca-fields (feat-gen "pca" "batchprojection")
+        dataset-assoc (assoc-feature-gen dataset-id  model-list name excluded)
+        all-fields [cluster-fields anomaly-fields topic-fields pca-fields]
+        all-ds (cons dataset-assoc (map batch-output-ds all-fields)))
+    (feature-generation-dataset all-ds name objective-id)))
+
+(define extended-datasets
+  (map (lambda(d) (when (= (resource-type d) "dataset")
+                    (log-info " - Extending dataset: " (resource-name d))
```
Comment (mmerce) [inline automl/automl-script/script.whizzml:56]: I guess is not very important, but jao asked me to indent things after when and if aligned with the condition parenthesis

File: automl/automl-script/script.whizzml [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+                        (create-dataset {"name" name
+                                         "origin_dataset" ds
+                                         "input_fields" fields}))))
+       extended-datasets))
+
+(log-featured "Model Selection")
+;; Creates the final fusion model with the top 5 models
+;; from an optiml generated with the filtered datasets
+(define output-fusion
+  (let ([train valid test] filtered-datasets)
```
Comment (mmerce) [inline automl/automl-script/script.whizzml]: you can replace "test" by "_" if you only need to use the train and valid info

File: automl/automl-script/script.whizzml [hunk 20]
```diff
@@ -0,0 +1,126 @@
...
+  (when (= (resource-type test-dataset) "dataset")
+    (let ([train valid test] filtered-datasets)
+      (prediction-from-fusion output-fusion test))))
+
+(log-info "Creating output evaluation")
+;; Creates the final evaluation from the validation
+;; dataset and the fusion
+(define output-evaluation
+  (when (= (resource-type validation-dataset) "dataset")
+    (let ([train valid test] filtered-datasets)
```
Comment (mmerce) [inline automl/automl-script/script.whizzml]: ditto: [_ valid _]

**Review by mmerce — COMMENTED** [review]
Basically nits and comments. I've not tested functionallity yet.

File: automl/automl-script/script.whizzml (hunk 17 above)
Comment (unmonoqueteclea) [reply automl/automl-script/script.whizzml:67]: Totally agree.

**Comment by unmonoqueteclea** [comment]
Thank you for your exhaustive review @mmerce 
It is great to improve my technical writing skills.

**Comment by mmerce** [comment]
There's a problem when no association rules are found

```
(define (_batch-association-sets dataset-id association-id  match . rule-ids)
  (let (association (fetch association-id)
        associations (association "associations" {})
        rule-ids (if (empty? (head rule-ids []))
                   (map (lambda (r) (r "id")) (associations "rules"))
```

it should be changed to (map (lambda (r) (r "id")) (associations "rules" []))

**Comment by mmerce** [comment]
In fact:

```
(define (_batch-association-sets dataset-id association-id  match . rule-ids)
  (let (association (fetch association-id)
        associations (association "associations" {})
        rules (associations "rules" [])
        rule-ids (if (empty? (head rule-ids []))
                   (map (lambda (r) (r "id")) rules)
                   (head rule-ids))
        rules (filter (lambda (r) (member? (r "id") rule-ids))
                      rules)
```

**Comment by mmerce** [comment]
And finally, in this case an error will be raised and should be caught 

```
    (when (empty? new-fields)
      (raise (str "No rules were found for the list of IDs: "
                  (join ", " rule-ids))))
```
Or probably, you could just remove the error and return an empty list of fields

**Comment by unmonoqueteclea** [comment]
> And finally, in this case an error will be raised and should be caught
> 
> ```
>     (when (empty? new-fields)
>       (raise (str "No rules were found for the list of IDs: "
>                   (join ", " rule-ids))))
> ```

catch the error or simply remove this raise?

**Comment by mmerce** [comment]
The script does not need it, you can remove it. The raise makes sense when the user provides a list of ids to filter the ones found in the association, because you need to check for mistakes. Here, there's no such a list.

**Review by mmerce — APPROVED** [review]
looks good

**Pull request merged**